lib/class-wp-rest-pattern-directory-controller.php

```diff
@@ -0,0 +1,262 @@
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to view the local pattern directory.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- Method must match signature of parent class.
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_pattern_directory_cannot_view',
+			__( 'Sorry, you are not allowed to browse the local block pattern directory.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Search and retrieve block patterns metadata
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$query_args   = array();
+		$category_ids = $request->get_param( 'category' );
+		$search_term  = $request->get_param( 'search' );
+
+		if ( $category_ids ) {
+			$query_args['pattern-categories'] = $category_ids;
+		}
+
+		if ( $search_term ) {
+			$query_args['search'] = $search_term;
+		}
+
+		$api_url = add_query_arg(
+			$query_args,
+			'http://api.wordpress.org/patterns/1.0/'
+		);
+
+		if ( wp_http_supports( array( 'ssl' ) ) ) {
+			$api_url = set_url_scheme( $api_url, 'https' );
+		}
+
+		$response     = wp_remote_get( $api_url );
+		$raw_patterns = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( is_wp_error( $response ) ) {
+			$response->add_data( array( 'status' => 500 ) );
+
+			return $response;
+		}
+
+		$prepared_patterns = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$prepared_patterns[] = $this->prepare_item_for_response( $pattern, $request );
+			}
+		}
+
+		return rest_ensure_response( $prepared_patterns );
+	}
+
+	/**
+	 * Prepare a raw pattern before it's output in an API response.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param object          $raw_pattern A pattern from api.wordpress.org, before any changes.
+	 * @param WP_REST_Request $request     Request object.
+	 *
+	 * @return array
+	 */
+	public function prepare_item_for_response( $raw_pattern, $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- Method must match signature of parent class.
```

Comment:
This should return a `WP_REST_Response` object per the documented return type. `get_items` should then be adjusted to call `prepare_response_for_collection()`.